acme/challenge.go

```diff
@@ -463,6 +501,333 @@ func deviceAttest01Validate(ctx context.Context, ch *Challenge, db DB, jwk *jose
+	if !ok {
+		return nil, NewError(ErrorBadAttestationStatementType, "x5c is malformed")
+	}
+	akCert, err := x509.ParseCertificate(akCertBytes)
+	if err != nil {
+		return nil, WrapError(ErrorBadAttestationStatementType, err, "x5c is malformed")
+	}
+
+	intermediates := x509.NewCertPool()
+	if len(x5c[1:]) > 0 {
+		for _, v := range x5c[1:] {
+			intCertBytes, vok := v.([]byte)
+			if !vok {
+				return nil, NewError(ErrorBadAttestationStatementType, "x5c is malformed")
+			}
+			intCert, err := x509.ParseCertificate(intCertBytes)
+			if err != nil {
+				return nil, WrapError(ErrorBadAttestationStatementType, err, "x5c is malformed")
+			}
+			intermediates.AddCert(intCert)
+		}
+	}
+
+	// TODO(hs): this can be removed when permanent-identifier/hardware-module-name are handled correctly in
+	// the stdlib in https://cs.opensource.google/go/go/+/refs/tags/go1.19:src/crypto/x509/parser.go;drc=b5b2cf519fe332891c165077f3723ee74932a647;l=362,
+	// but I doubt that will happen.
+	if len(akCert.UnhandledCriticalExtensions) > 0 {
+		unhandledCriticalExtensions := akCert.UnhandledCriticalExtensions[:0]
+		for _, extOID := range akCert.UnhandledCriticalExtensions {
+			switch {
+			case extOID.Equal(oidSubjectAlternativeName):
+				// allow Subject Alternative Names, including PermanentIdentifier, HardwareModuleName, TPM attributes, etc
+			default:
+				// OIDs that are not in the switch with explicitly allowed OIDs remain unhandled
+				unhandledCriticalExtensions = append(unhandledCriticalExtensions, extOID)
+			}
+		}
+		akCert.UnhandledCriticalExtensions = unhandledCriticalExtensions
+	}
+
+	roots, ok := prov.GetAttestationRoots()
+	if !ok {
+		return nil, NewErrorISE("failed getting TPM attestation root CAs")
+	}
+
+	verifiedChains, err := akCert.Verify(x509.VerifyOptions{
+		Roots:         roots,
+		Intermediates: intermediates,
+		CurrentTime:   time.Now().Truncate(time.Second),
+		KeyUsages:     []x509.ExtKeyUsage{x509.ExtKeyUsageAny},
+	})
+	if err != nil {
+		return nil, WrapError(ErrorBadAttestationStatementType, err, "x5c is not valid")
+	}
+
+	// TODO(hs): implement revocation check; Verify() doesn't perform CRL check nor OCSP lookup.
+
+	// extract and validate Subject Alternative Name extension to contain at least one PermanentIdentifier
+	var sanExtension pkix.Extension
+	for _, ext := range akCert.Extensions {
+		if ext.Id.Equal(oidSubjectAlternativeName) {
+			sanExtension = ext
+			break
+		}
+	}
+
+	if sanExtension.Value == nil {
+		return nil, NewError(ErrorBadAttestationStatementType, "AK certificate is missing Subject Alternative Name extension")
+	}
+
+	sans, err := parseSANs(sanExtension)
+	if err != nil {
+		return nil, WrapError(ErrorBadAttestationStatementType, err, "failed parsing AK certificate SAN extension")
+	}
+
+	var permanentIdentifiers []string
+	for _, san := range sans {
+		if san.Type == x509util.PermanentIdentifierType {
+			permanentIdentifiers = append(permanentIdentifiers, san.Value)
+		}
+	}
```

Comment:
All this will be something like:
```go
sans, err := x509util.ParseSubjectAlternativeNames(cert)
if err != nil {
   return nil, WrapError(Foo, err, "whatever")
}
permanentIdentifiers := sans.PermanentIdentifiers
```

Reply:
I've added the function in https://github.com/smallstep/crypto/pull/209. It's currently on top of the `tpm` PR, but I can rebase it if you want. Is this going into the right direction? I think it makes sense to copy of the DNS and the other known SANs too, to get the complete set of SANs. Taken strictly it's not parsing those, so maybe `Extract...` covers the functionality better?